x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceServiceTests.java

```diff
@@ -167,24 +186,63 @@ public void testGetSessionDoesNotLeakFileIfClosed() throws IOException {
             indexDoc(indexShard, "_doc", Integer.toString(i));
             flushShard(indexShard, true);
         }
-        final String sessionUUID1 = UUIDs.randomBase64UUID();
+        final String sessionUUID = UUIDs.randomBase64UUID();
 
-        restoreSourceService.openSession(sessionUUID1, indexShard);
+        restoreSourceService.openSession(sessionUUID, indexShard);
 
         ArrayList<StoreFileMetaData> files = new ArrayList<>();
         indexShard.snapshotStoreMetadata().forEach(files::add);
 
-        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID1)) {
+        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID)) {
             sessionReader.readFileBytes(files.get(0).name(), new BytesArray(new byte[10]));
         }
 
         // Request a second file to ensure that original file is not leaked
-        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID1)) {
+        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID)) {
             sessionReader.readFileBytes(files.get(1).name(), new BytesArray(new byte[10]));
         }
 
-        restoreSourceService.closeSession(sessionUUID1);
+        restoreSourceService.closeSession(sessionUUID);
         closeShards(indexShard);
         // Exception will be thrown if file is not closed.
     }
+
+    public void testSessionCanTimeout() throws Exception {
+        IndexShard indexShard = newStartedShard(true);
+
+        final String sessionUUID = UUIDs.randomBase64UUID();
+
+        restoreSourceService.openSession(sessionUUID, indexShard);
+
+        // Session starts as not idle. First task will mark it as idle
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+        // Task is still scheduled
+        assertTrue(taskQueue.hasDeferredTasks());
+
+        // Accessing session marks it as not-idle
+        try (CcrRestoreSourceService.SessionReader reader = restoreSourceService.getSessionReader(sessionUUID)) {
+            // Check session exists
+        }
+
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+        // Task is still scheduled
+        assertTrue(taskQueue.hasDeferredTasks());
+
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+        // Task is cancelled when the session times out
+        assertFalse(taskQueue.hasDeferredTasks());
+
+        try (CcrRestoreSourceService.SessionReader reader = restoreSourceService.getSessionReader(sessionUUID)) {
+            fail("Should have timed out.");
```

Comment:
this can be written more concisely with `expectThrows`. Can you also assert something on the exception message?